queasars/job_shop_scheduling/problem_instances.py

```diff
@@ -0,0 +1,375 @@
+        self._schedule: dict[Job, tuple[ScheduledOperation, ...]] = schedule
+        self._is_valid: Optional[bool] = None
+        self._makespan: Optional[int] = None
+
+    @property
+    def problem_instance(self) -> JobShopSchedulingProblemInstance:
+        """
+        :return: the problem instance for which this result represents an attempted solution
+        :rtype: JobShopSchedulingProblemInstance
+        """
+        return self._problem_instance
+
+    @property
+    def schedule(self) -> dict[Job, tuple[ScheduledOperation, ...]]:
+        """
+        :return: schedule which attempts to solve the given problem instance
+        :rtype: dict[Job, tuple[ScheduledOperation, ...]]
+        """
+        return self._schedule
+
+    @property
+    def is_valid(self) -> bool:
+        """
+        :return: true, if the result adheres to all constraints of the JSSP and false otherwise
+        :rtype: bool
+        """
+        if self._is_valid is not None:
+            return self._is_valid
+        is_valid = self._is_valid_solution()
+        self._is_valid = is_valid
+        return is_valid
+
+    @property
+    def makespan(self) -> Optional[int]:
+        """
+        :return: the integer makespan if the result is valid and None otherwise
+        :rtype: Optional[int]
+        """
+        if not self.is_valid:
+            return None
+        if self._makespan is not None:
+            return self._makespan
+        makespan: int = max((scheduled_operations[-1].end_time for scheduled_operations in self.schedule.values()))
+        self._makespan = makespan
+        return makespan
+
+    def _is_valid_solution(self) -> bool:
+        """
+        Checks whether this JobShopSchedulingResult is a valid solution. Therefore, it checks
+        that all operations are in the correct order and that no operation per job and per machine may
+        overlap
+
+        :return: true if the solution is valid and false otherwise
+        :rtype: bool
+        """
+        machine_operation_mapping: dict[Machine, list[ScheduledOperation]] = {
+            machine: [] for machine in self._problem_instance.machines
+        }
+
+        # Check that all operations per job are correctly ordered and not overlapping
+        for job in self._problem_instance.jobs:
+            previous_scheduled_operation: Optional[ScheduledOperation] = None
+            for scheduled_operation in self._schedule[job]:
+                machine_operation_mapping[scheduled_operation.operation.machine].append(scheduled_operation)
+                # All scheduled operations are checked here before they are accessed.
+                if not scheduled_operation.is_scheduled:
+                    return False
+                if previous_scheduled_operation is not None:
+                    if scheduled_operation.start_time < previous_scheduled_operation.end_time:
+                        return False
+                previous_scheduled_operation = scheduled_operation
```

Comment:
I guess this can be simplified if you can assume the input (or the field after the constructor) is already ordered (according to start time or similar)?

Reply:
You do not want to order them according to start time, as the order within the job is alread fixed. You cannot alter it. The only check that needs to be made, is that the scheduled times actually adhere to it.